diff --git a/CMakeLists.txt b/CMakeLists.txt
index ec978fb..c8e29c4 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -19,7 +19,8 @@ file(GLOB SRCS_EMULATOR
         Libs/WouoFix.cpp
         *.c 
         *.cpp
-        src/Arduino.c
+        src/Arduino.cpp
+        WouoUI-Demo/ui-128_64/*.cpp
         )
 include_directories(Libs/u8g2/csrc/)
 include_directories(Libs/u8g2/cppsrc/)
diff --git a/Libs/WouoFix.h b/Libs/WouoFix.h
index a0f701f..e57513d 100644
--- a/Libs/WouoFix.h
+++ b/Libs/WouoFix.h
@@ -1,3 +1,8 @@
+#ifndef WOUOFIX_H
+#define WOUOFIX_H
+
+#include "Arduino/ArduinoCompatible.h"
+
 #define KEY_ESC 0
 #define KEY_F1 0
 #define KEY_F2 0
@@ -62,3 +67,4 @@
 void ui_param_init();
 
 
+#endif
\ No newline at end of file
diff --git a/WouoUI-Demo/Wouo.h b/WouoUI-Demo/Wouo.h
index 671f672..12f07b1 100644
--- a/WouoUI-Demo/Wouo.h
+++ b/WouoUI-Demo/Wouo.h
@@ -1,7 +1,10 @@
 
-
-
-
+#ifndef WOUO_H
+#define WOUO_H
+#include "WouoFix.h"
+typedef unsigned char uint8_t;
+typedef unsigned short uint16_t;
+typedef signed short int16_t;
 /************************************* 定义页面 *************************************/
 
 // 总目录，缩进表示页面层级
@@ -29,6 +32,20 @@ enum
     S_NONE,      // 直接选择页面
 };
 
+typedef struct
+{
+    uint8_t id;
+    bool flag;
+    bool pressed;
+    bool CW_1;
+    bool CW_2;
+    bool val;
+    bool val_last;
+    bool alv;
+    bool blv;
+    long count;
+} InputContext;
+
 // 菜单结构体
 typedef struct MENU
 {
@@ -88,8 +105,7 @@ const uint8_t main_icon_pic[][120] = {
 // OLED变量
 #define DISP_H 64  // 屏幕高度
 #define DISP_W 128 // 屏幕宽度
-uint8_t *buf_ptr;  // 指向屏幕缓冲的指针
-uint16_t buf_len;  // 缓冲长度
+
 
 // UI变量
 #define UI_DEPTH 20  // 最深层级数
@@ -158,8 +174,7 @@ struct
 } list;
 
 // 电压测量页面变量
-// 开发板模拟引脚
-uint8_t analog_pin[10] = {PA0, PA1, PA2, PA3, PA4, PA5, PA6, PA7, PB0, PB1};
+
 // 曲线相关
 #define WAVE_SAMPLE 20  // 采集倍数
 #define WAVE_W 94       // 波形宽度
@@ -283,172 +298,6 @@ typedef struct
     uint8_t fade;
     uint8_t param[UI_PARAM];
 } UiContext;
-/************************************* 文字内容 *************************************/
-
-M_SELECT main_menu[]{
-    {"Sleep"},
-    {"Editor"},
-    {"Volt"},
-    {"Setting"},
-};
-
-M_SELECT editor_menu[]{
-    {"[ Editor ]"},
-    {"- Function 0"},
-    {"- Function 1"},
-    {"- Function 2"},
-    {"- Function 3"},
-    {"- Function 4"},
-    {"- Function 5"},
-    {"- Function 6"},
-    {"- Function 7"},
-    {"- Function 8"},
-    {"- Function 9"},
-    {"- Knob"},
-};
-
-M_SELECT knob_menu[]{
-    {"[ Knob ]"},
-    {"# Rotate Func"},
-    {"$ Press Func"},
-};
 
-M_SELECT krf_menu[]{
-    {"[ Rotate Function ]"},
-    {"--------------------------"},
-    {"= Disable"},
-    {"--------------------------"},
-    {"= Volume"},
-    {"= Brightness"},
-    {"--------------------------"},
-};
-
-M_SELECT kpf_menu[]{
-    {"[ Press Function ]"},
-    {"--------------------------"},
-    {"= Disable"},
-    {"--------------------------"},
-    {"= A"},
-    {"= B"},
-    {"= C"},
-    {"= D"},
-    {"= E"},
-    {"= F"},
-    {"= G"},
-    {"= H"},
-    {"= I"},
-    {"= J"},
-    {"= K"},
-    {"= L"},
-    {"= M"},
-    {"= N"},
-    {"= O"},
-    {"= P"},
-    {"= Q"},
-    {"= R"},
-    {"= S"},
-    {"= T"},
-    {"= U"},
-    {"= V"},
-    {"= W"},
-    {"= X"},
-    {"= Y"},
-    {"= Z"},
-    {"--------------------------"},
-    {"= 0"},
-    {"= 1"},
-    {"= 2"},
-    {"= 3"},
-    {"= 4"},
-    {"= 5"},
-    {"= 6"},
-    {"= 7"},
-    {"= 8"},
-    {"= 9"},
-    {"--------------------------"},
-    {"= Esc"},
-    {"= F1"},
-    {"= F2"},
-    {"= F3"},
-    {"= F4"},
-    {"= F5"},
-    {"= F6"},
-    {"= F7"},
-    {"= F8"},
-    {"= F9"},
-    {"= F10"},
-    {"= F11"},
-    {"= F12"},
-    {"--------------------------"},
-    {"= Left Ctrl"},
-    {"= Left Shift"},
-    {"= Left Alt"},
-    {"= Left Win"},
-    {"= Right Ctrl"},
-    {"= Right Shift"},
-    {"= Right Alt"},
-    {"= Right Win"},
-    {"--------------------------"},
-    {"= Caps Lock"},
-    {"= Backspace"},
-    {"= Return"},
-    {"= Insert"},
-    {"= Delete"},
-    {"= Tab"},
-    {"--------------------------"},
-    {"= Home"},
-    {"= End"},
-    {"= Page Up"},
-    {"= Page Down"},
-    {"--------------------------"},
-    {"= Up Arrow"},
-    {"= Down Arrow"},
-    {"= Left Arrow"},
-    {"= Right Arrow"},
-    {"--------------------------"},
-};
 
-M_SELECT volt_menu[]{
-    {"A0"},
-    {"A1"},
-    {"A2"},
-    {"A3"},
-    {"A4"},
-    {"A5"},
-    {"A6"},
-    {"A7"},
-    {"B0"},
-    {"B1"},
-};
-
-M_SELECT setting_menu[]{
-    {"[ Setting ]"},
-    {"~ Disp Bri"},
-    {"~ Tile Ani"},
-    {"~ List Ani"},
-    {"~ Win Ani"},
-    {"~ Spot Ani"},
-    {"~ Tag Ani"},
-    {"~ Fade Ani"},
-    {"~ Btn SPT"},
-    {"~ Btn LPT"},
-    {"+ T Ufd Fm Scr"},
-    {"+ L Ufd Fm Scr"},
-    {"+ T Loop Mode"},
-    {"+ L Loop Mode"},
-    {"+ Win Bokeh Bg"},
-    {"+ Knob Rot Dir"},
-    {"+ Dark Mode"},
-    {"- [ About ]"},
-};
-
-M_SELECT about_menu[]{
-    {"[ WouoUI ]"},
-    {"- Version: v2.3"},
-    {"- Board: STM32F103"},
-    {"- Ram: 20k"},
-    {"- Flash: 64k"},
-    {"- Freq: 72Mhz"},
-    {"- Creator: RQNG"},
-    {"- Bili UID: 9182439"},
-};
+#endif
diff --git a/WouoUI-Demo/ui-128_64/WouoUI-128_64.cpp b/WouoUI-Demo/ui-128_64/WouoUI-128_64.cpp
index 637ab31..a545ebe 100644
--- a/WouoUI-Demo/ui-128_64/WouoUI-128_64.cpp
+++ b/WouoUI-Demo/ui-128_64/WouoUI-128_64.cpp
@@ -56,6 +56,9 @@
 #include <monitor.h>
 
 #include "../Wouo.h"
+#include "WouoUI_init.h"
+#include "WouoUI_proc.h"
+#include "WouoUI-128_64.h"
 
 #define SCL PB6
 #define SDA PB7
@@ -67,6 +70,180 @@ U8G2_SSD1306_128X64_NONAME_F_HW_I2C u8g2(U8G2_R0, SCL, SDA, RST); // 分辨率
 UiContext ui;
 ParamWindow win;
 
+/************************************* 文字内容 *************************************/
+
+M_SELECT main_menu[]{
+    {"Sleep"},
+    {"Editor"},
+    {"Volt"},
+    {"Setting"},
+};
+
+M_SELECT editor_menu[]{
+    {"[ Editor ]"},
+    {"- Function 0"},
+    {"- Function 1"},
+    {"- Function 2"},
+    {"- Function 3"},
+    {"- Function 4"},
+    {"- Function 5"},
+    {"- Function 6"},
+    {"- Function 7"},
+    {"- Function 8"},
+    {"- Function 9"},
+    {"- Knob"},
+};
+
+M_SELECT knob_menu[]{
+    {"[ Knob ]"},
+    {"# Rotate Func"},
+    {"$ Press Func"},
+};
+
+M_SELECT krf_menu[]{
+    {"[ Rotate Function ]"},
+    {"--------------------------"},
+    {"= Disable"},
+    {"--------------------------"},
+    {"= Volume"},
+    {"= Brightness"},
+    {"--------------------------"},
+};
+
+M_SELECT kpf_menu[]{
+    {"[ Press Function ]"},
+    {"--------------------------"},
+    {"= Disable"},
+    {"--------------------------"},
+    {"= A"},
+    {"= B"},
+    {"= C"},
+    {"= D"},
+    {"= E"},
+    {"= F"},
+    {"= G"},
+    {"= H"},
+    {"= I"},
+    {"= J"},
+    {"= K"},
+    {"= L"},
+    {"= M"},
+    {"= N"},
+    {"= O"},
+    {"= P"},
+    {"= Q"},
+    {"= R"},
+    {"= S"},
+    {"= T"},
+    {"= U"},
+    {"= V"},
+    {"= W"},
+    {"= X"},
+    {"= Y"},
+    {"= Z"},
+    {"--------------------------"},
+    {"= 0"},
+    {"= 1"},
+    {"= 2"},
+    {"= 3"},
+    {"= 4"},
+    {"= 5"},
+    {"= 6"},
+    {"= 7"},
+    {"= 8"},
+    {"= 9"},
+    {"--------------------------"},
+    {"= Esc"},
+    {"= F1"},
+    {"= F2"},
+    {"= F3"},
+    {"= F4"},
+    {"= F5"},
+    {"= F6"},
+    {"= F7"},
+    {"= F8"},
+    {"= F9"},
+    {"= F10"},
+    {"= F11"},
+    {"= F12"},
+    {"--------------------------"},
+    {"= Left Ctrl"},
+    {"= Left Shift"},
+    {"= Left Alt"},
+    {"= Left Win"},
+    {"= Right Ctrl"},
+    {"= Right Shift"},
+    {"= Right Alt"},
+    {"= Right Win"},
+    {"--------------------------"},
+    {"= Caps Lock"},
+    {"= Backspace"},
+    {"= Return"},
+    {"= Insert"},
+    {"= Delete"},
+    {"= Tab"},
+    {"--------------------------"},
+    {"= Home"},
+    {"= End"},
+    {"= Page Up"},
+    {"= Page Down"},
+    {"--------------------------"},
+    {"= Up Arrow"},
+    {"= Down Arrow"},
+    {"= Left Arrow"},
+    {"= Right Arrow"},
+    {"--------------------------"},
+};
+
+M_SELECT volt_menu[]{
+    {"A0"},
+    {"A1"},
+    {"A2"},
+    {"A3"},
+    {"A4"},
+    {"A5"},
+    {"A6"},
+    {"A7"},
+    {"B0"},
+    {"B1"},
+};
+
+M_SELECT setting_menu[]{
+    {"[ Setting ]"},
+    {"~ Disp Bri"},
+    {"~ Tile Ani"},
+    {"~ List Ani"},
+    {"~ Win Ani"},
+    {"~ Spot Ani"},
+    {"~ Tag Ani"},
+    {"~ Fade Ani"},
+    {"~ Btn SPT"},
+    {"~ Btn LPT"},
+    {"+ T Ufd Fm Scr"},
+    {"+ L Ufd Fm Scr"},
+    {"+ T Loop Mode"},
+    {"+ L Loop Mode"},
+    {"+ Win Bokeh Bg"},
+    {"+ Knob Rot Dir"},
+    {"+ Dark Mode"},
+    {"- [ About ]"},
+};
+
+M_SELECT about_menu[]{
+    {"[ WouoUI ]"},
+    {"- Version: v2.3"},
+    {"- Board: STM32F103"},
+    {"- Ram: 20k"},
+    {"- Flash: 64k"},
+    {"- Freq: 72Mhz"},
+    {"- Creator: RQNG"},
+    {"- Bili UID: 9182439"},
+};
+
+uint8_t *buf_ptr;  // 指向屏幕缓冲的指针
+uint16_t buf_len;  // 缓冲长度
+// 开发板模拟引脚
+uint8_t analog_pin[10] = {PA0, PA1, PA2, PA3, PA4, PA5, PA6, PA7, PB0, PB1};
 /************************************* USB 相关 *************************************/
 
 #include <USBComposite.h>
@@ -165,27 +342,13 @@ void eeprom_init()
 #define BIO PB13
 #define SW PB14
 
-// 按键ID
-#define BTN_ID_CC 0 // 逆时针旋转
-#define BTN_ID_CW 1 // 顺时针旋转
-#define BTN_ID_SP 2 // 短按
-#define BTN_ID_LP 3 // 长按
+
 
 // 按键变量
 #define BTN_PARAM_TIMES 2 // 由于uint8_t最大值可能不够，但它存储起来方便，这里放大两倍使用
-struct
-{
-    uint8_t id;
-    bool flag;
-    bool pressed;
-    bool CW_1;
-    bool CW_2;
-    bool val;
-    bool val_last;
-    bool alv;
-    bool blv;
-    long count;
-} volatile btn;
+
+
+InputContext btn;
 
 void knob_inter()
 {
@@ -301,26 +464,7 @@ void window_value_init(char title[], uint8_t select, uint8_t *value, uint8_t max
 
 /*********************************** UI 初始化函数 *********************************/
 
-// 在初始化EEPROM时，选择性初始化的默认设置
-void ui_param_init()
-{
-    ui.param[DISP_BRI] = 255; // 屏幕亮度
-    ui.param[TILE_ANI] = 30;  // 磁贴动画速度
-    ui.param[LIST_ANI] = 60;  // 列表动画速度
-    ui.param[WIN_ANI] = 25;   // 弹窗动画速度
-    ui.param[SPOT_ANI] = 50;  // 聚光动画速度
-    ui.param[TAG_ANI] = 60;   // 标签动画速度
-    ui.param[FADE_ANI] = 30;  // 消失动画速度
-    ui.param[BTN_SPT] = 25;   // 按键短按时长
-    ui.param[BTN_LPT] = 150;  // 按键长按时长
-    ui.param[TILE_UFD] = 1;   // 磁贴图标从头展开开关
-    ui.param[LIST_UFD] = 1;   // 菜单列表从头展开开关
-    ui.param[TILE_LOOP] = 0;  // 磁贴图标循环模式开关
-    ui.param[LIST_LOOP] = 0;  // 菜单列表循环模式开关
-    ui.param[WIN_BOK] = 0;    // 弹窗背景虚化开关
-    ui.param[KNOB_DIR] = 0;   // 旋钮方向切换开关
-    ui.param[DARK_MODE] = 1;  // 黑暗模式开关
-}
+
 
 // 列表类页面列表行数初始化，必须初始化的参数
 void ui_init()
@@ -1061,48 +1205,6 @@ void sleep_proc()
     }
 }
 
-// 主菜单处理函数，磁贴类模板
-void main_proc()
-{
-    tile_show(main_menu, main_icon_pic);
-    if (btn.pressed)
-    {
-        btn.pressed = false;
-        switch (btn.id)
-        {
-        case BTN_ID_CW:
-        case BTN_ID_CC:
-            tile_rotate_switch();
-            break;
-        case BTN_ID_SP:
-            switch (ui.select[ui.layer])
-            {
-
-            case 0:
-                ui.index = M_SLEEP;
-                ui.state = S_LAYER_OUT;
-                break;
-            case 1:
-                ui.index = M_EDITOR;
-                ui.state = S_LAYER_IN;
-                break;
-            case 2:
-                ui.index = M_VOLT;
-                ui.state = S_LAYER_IN;
-                break;
-            case 3:
-                ui.index = M_SETTING;
-                ui.state = S_LAYER_IN;
-                break;
-            }
-        }
-        if (!tile.select_flag && ui.init)
-        {
-            tile.indi_x = 0;
-            tile.title_y = tile.title_y_calc;
-        }
-    }
-}
 
 // 编辑器菜单处理函数
 void editor_proc()
diff --git a/WouoUI-Demo/ui-128_64/WouoUI_proc.cpp b/WouoUI-Demo/ui-128_64/WouoUI_proc.cpp
index e69de29..f186677 100644
--- a/WouoUI-Demo/ui-128_64/WouoUI_proc.cpp
+++ b/WouoUI-Demo/ui-128_64/WouoUI_proc.cpp
@@ -0,0 +1,50 @@
+#include "WouoUI_proc.h"
+#include "../Wouo.h"
+#include "WouoUI-128_64.h"
+
+extern M_SELECT main_menu[];
+extern InputContext btn;
+extern UiContext ui;
+
+// 主菜单处理函数，磁贴类模板
+void main_proc()
+{
+    tile_show(main_menu, main_icon_pic);
+    if (btn.pressed)
+    {
+        btn.pressed = false;
+        switch (btn.id)
+        {
+        case BTN_ID_CW:
+        case BTN_ID_CC:
+            tile_rotate_switch();
+            break;
+        case BTN_ID_SP:
+            switch (ui.select[ui.layer])
+            {
+
+            case 0:
+                ui.index = M_SLEEP;
+                ui.state = S_LAYER_OUT;
+                break;
+            case 1:
+                ui.index = M_EDITOR;
+                ui.state = S_LAYER_IN;
+                break;
+            case 2:
+                ui.index = M_VOLT;
+                ui.state = S_LAYER_IN;
+                break;
+            case 3:
+                ui.index = M_SETTING;
+                ui.state = S_LAYER_IN;
+                break;
+            }
+        }
+        if (!tile.select_flag && ui.init)
+        {
+            tile.indi_x = 0;
+            tile.title_y = tile.title_y_calc;
+        }
+    }
+}
\ No newline at end of file
diff --git a/WouoUI-Demo/ui-128_64/WouoUI_proc.h b/WouoUI-Demo/ui-128_64/WouoUI_proc.h
index e69de29..087119a 100644
--- a/WouoUI-Demo/ui-128_64/WouoUI_proc.h
+++ b/WouoUI-Demo/ui-128_64/WouoUI_proc.h
@@ -0,0 +1,7 @@
+#ifndef WOUOUI_PROC_H
+#define WOUOUI_PROC_H
+
+void main_proc();
+
+
+#endif /* WOUOUI_PROC_H */
\ No newline at end of file
diff --git a/main.cpp b/main.cpp
index 7641620..d76fa66 100644
--- a/main.cpp
+++ b/main.cpp
@@ -3,6 +3,6 @@
 // #include "../../WouoUI-128_64/WouoUI-128_64.ino"
 // #include "../../WouoUI-128_128/WouoUI-128_128.ino"
 // #include "../../WouoUI-Lite_General/WouoUI-Lite_General.ino"
-#include "WouoUI-Demo/ui-128_64/WouoUI-128_64.cpp"
+// #include "WouoUI-Demo/ui-128_64/WouoUI-128_64.cpp"
 
 // export DISPLAY=:0
\ No newline at end of file
diff --git a/src/Arduino.c b/src/Arduino.c
deleted file mode 100644
index f307015..0000000
--- a/src/Arduino.c
+++ /dev/null
@@ -1,26 +0,0 @@
-
-#include "u8g2.h"
-#include <stdio.h>
-#define SDL_MAIN_HANDLED
-#include <SDL2/SDL.h>
-#include "Arduino.h"
-#include <stdbool.h>
-#include <monitor.h>
-extern volatile bool sdl_quit_qry;
-
-void postLoop();
-
-int main(int argc, char *argv[])
-{
-    printf("pro in\n");
-    setup();
-    while (1){
-        loop();
-        // postLoop();
-        SDL_Delay(5);
-
-    }
-    printf("pro out\n");
-    return 1;
-}
-
diff --git a/src/Arduino.h b/src/Arduino.h
index c43a990..a1d7b03 100644
--- a/src/Arduino.h
+++ b/src/Arduino.h
@@ -1,13 +1,13 @@
-#ifdef __cplusplus
-#include "ArduinoCompatible.h"
-extern "C"
-{
-#endif
+// #ifdef __cplusplus
+// #include "ArduinoCompatible.h"
+// extern "C"
+// {
+// #endif
 
-void setup();
-void loop();
+// void setup();
+// void loop();
 
 
-#ifdef __cplusplus
-}
-#endif
\ No newline at end of file
+// #ifdef __cplusplus
+// }
+// #endif
\ No newline at end of file
